tracer/src/Datadog.Trace/AppSec/BodyExtractor.cs

```diff
@@ -0,0 +1,180 @@
+// <copyright file="BodyExtractor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using Datadog.Trace.AppSec.Waf;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.AppSec
+{
+    internal static class BodyExtractor
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(BodyExtractor));
+
+        private static readonly Regex NameExtractor = new Regex(@"\<(?'PropertyName'[^\>]+)\>", RegexOptions.Compiled);
+
+        private static readonly HashSet<Type> AdditionalPrimitives = new()
+        {
+            typeof(string),
+            typeof(decimal),
+            typeof(Guid),
+            typeof(DateTime),
+            typeof(DateTimeOffset),
+            typeof(TimeSpan)
+        };
+
+        internal static object GetKeysAndValues(object body)
+        {
+            var visted = new HashSet<object>();
+            var item = ExtractType(body.GetType(), body, 0, visted);
+
+            return item;
+        }
+
+        private static bool IsOurKindOfPrimitive(Type t)
+        {
+            return t.IsPrimitive || AdditionalPrimitives.Contains(t) || t.IsEnum;
+        }
+
+        private static void ExtractProperties(Dictionary<string, object> dic, object body, int depth, HashSet<object> visited)
+        {
+            if (visited.Contains(body))
+            {
+                return;
+            }
+
+            visited.Add(body);
+
+            Log.Debug("ExtractProperties - body: {Body}", body);
+
+            var fields =
+                    body.GetType()
+                        .GetFields(BindingFlags.Instance | BindingFlags.NonPublic)
+                        .Where(x => x.IsPrivate && x.Name.EndsWith("__BackingField"))
+                        .ToArray();
+
+            depth++;
+
+            for (var i = 0; i < fields.Length; i++)
+            {
+                if (dic.Count >= WafConstants.MaxMapOrArrayLength || depth >= WafConstants.MaxObjectDepth)
+                {
+                    return;
+                }
+
+                var field = fields[i];
+
+                var nameMatch = NameExtractor.Match(field.Name);
+                if (!nameMatch.Success || nameMatch.Groups.Count == 0)
+                {
+                    throw new Exception("Can't extract property name from " + field.Name);
```

Comment:
I implemented this as a throw, as it would have been quite weird if we failed to find a match, so my thinking was to stop further processing probably wouldn't be useful.

However, a badly formed field name could be a one off, so it might be worth trying to continue processing. I implemented this in: https://github.com/DataDog/dd-trace-dotnet/commit/dd3e5c55480f887f136eb420c27efcc15917be74